### PR TITLE
Improve benchmarks

### PR DESCRIPTION
**Pull request by sirenkovladd**
Add async benchmarks
use process.hrtime()

#### Checklist

- [X] run `npm run test` and `npm run benchmark`
- [X] tests and/or benchmarks are included
- [X] documentation is changed or added
- [X] commit message and code follows the [Developer's Certification of Origin](https://github.com/fastify/.github/blob/master/CONTRIBUTING.md#developers-certificate-of-origin-11)
      and the [Code of conduct](https://github.com/fastify/.github/blob/master/CODE_OF_CONDUCT.md)

### CODE REVIEW

**Comment by sirenkovladd** [comment]
master
```sh
❯ node ./benchmark/benchmark.js
Request x 331,715 ops/sec ±4.20% (81 runs sampled)
Custom Request x 25,142 ops/sec ±2.08% (85 runs sampled)
Request With Cookies x 232,410 ops/sec ±6.46% (82 runs sampled)
Request With Cookies n payload x 213,823 ops/sec ±3.60% (81 runs sampled)
ParseUrl x 1,313,090 ops/sec ±1.78% (89 runs sampled)
ParseUrl and query x 115,189 ops/sec ±1.75% (89 runs sampled)
```

fix-banchmark
```sh
Request x 345031 ops/sec ±15.56% (1725172 runs sampled)
Custom Request x 27734 ops/sec ±7.82% (138679 runs sampled)
Request With Cookies x 246922 ops/sec ±7.64% (1234610 runs sampled)
Request With Cookies n payload x 216127 ops/sec ±9.23% (1080638 runs sampled)
ParseUrl x 922286 ops/sec ±12.55% (4611434 runs sampled)
ParseUrl and query x 110163 ops/sec ±5.72% (550814 runs sampled)
read request body JSON x 95963 ops/sec ±12.21% (479814 runs sampled)
read request body buffer x 113153 ops/sec ±9.63% (565767 runs sampled)
read request body readable x 57184 ops/sec ±19.42% (285924 runs sampled)
Response write end x 29143 ops/sec ±65.02% (145726 runs sampled)
Response writeHead end x 25164 ops/sec ±197.25% (125837 runs sampled)
```

**Review by Uzlopak — CHANGES_REQUESTED** [review]
i am -1 of a self implemented benchmark.js

**Comment by Uzlopak** [comment]
use alternatively tinybench, if you dislike benchmark.js. it is also used by vitest.

**Comment by sirenkovladd** [comment]
Cool, I'll switch to a tinybench if you don't mind)

**Comment by sirenkovladd** [comment]
```sh
❯ node ./benchmark/benchmark.js
┌─────────┬──────────────────────────────────┬─────────────┬────────────────────┬───────────┬─────────┐
│ (index) │            Task Name             │   ops/sec   │ Average Time (ns)  │  Margin   │ Samples │
├─────────┼──────────────────────────────────┼─────────────┼────────────────────┼───────────┼─────────┤
│    0    │            'Request'             │  '343,888'  │ 2907.9191370450485 │ '±1.92%'  │ 171945  │
│    1    │         'Custom Request'         │  '28,236'   │ 35414.67135123085  │ '±2.08%'  │  14120  │
│    2    │      'Request With Cookies'      │  '256,366'  │ 3900.6658665664277 │ '±1.75%'  │ 128184  │
│    3    │ 'Request With Cookies n payload' │  '215,695'  │ 4636.170700281984  │ '±2.07%'  │ 107848  │
│    4    │            'ParseUrl'            │ '1,062,369' │ 941.2924302962186  │ '±1.25%'  │ 531185  │
│    5    │       'ParseUrl and query'       │  '119,659'  │ 8357.076274398258  │ '±0.80%'  │  59830  │
│    6    │     'read request body JSON'     │  '84,748'   │ 11799.619041712936 │ '±1.48%'  │  42375  │
│    7    │    'read request body buffer'    │  '114,046'  │ 8768.364248272579  │ '±1.58%'  │  57024  │
│    8    │   'read request body readable'   │  '63,504'   │ 15746.902949084779 │ '±3.14%'  │  31753  │
│    9    │       'Response write end'       │  '27,661'   │  36151.2651495023  │ '±13.34%' │  13831  │
│   10    │     'Response writeHead end'     │  '35,904'   │  27851.3512740605  │ '±6.76%'  │  17953  │
│   11    │          'base inject'           │  '18,365'   │ 54450.594459135114 │ '±13.36%' │  9183   │
└─────────┴──────────────────────────────────┴─────────────┴────────────────────┴───────────┴─────────┘
```

**Review by Uzlopak — APPROVED** [review]
I approve, but prefer that a core maintainer OKs the switch to tinybench.

File: benchmark/benchmark.js [hunk 1]
```diff
@@ -77,7 +103,62 @@ suite
...
+    await new Promise((resolve) => {
+      const res = new Response(req, () => resolve())
+      res.write('foo')
+      res.end()
+    })
+  })
+  .add('Response writeHead end', async function () {
+    const req = new Request(mockReq)
+    await new Promise((resolve) => {
+      const res = new Response(req, () => resolve())
```
Comment (Uzlopak) [inline benchmark/benchmark.js:144]:
```suggestion
      const res = new Response(req, resolve)
```

File: benchmark/benchmark.js [hunk 2]
```diff
@@ -77,7 +103,62 @@ suite
...
+      req.prepare(() => {
+        req.on('data', () => {})
+        req.on('end', resolve)
+      })
+    })
+  })
+  .add('Response write end', async function () {
+    const req = new Request(mockReq)
+    await new Promise((resolve) => {
+      const res = new Response(req, () => resolve())
```
Comment (Uzlopak) [inline benchmark/benchmark.js:136]:
```suggestion
      const res = new Response(req, resolve)
```

File: benchmark/benchmark.js [hunk 3]
```diff
@@ -77,7 +103,62 @@ suite
       xs: ['foo', 'bar']
     })
   })
-  .on('cycle', function (event) {
-    console.log(String(event.target))
+  .add('read request body JSON', async function () {
```
Comment (mcollina) [inline benchmark/benchmark.js:106]: remove async

File: benchmark/benchmark.js [hunk 4]
```diff
@@ -77,7 +103,62 @@ suite
...
+    await new Promise((resolve) => {
```
Comment (mcollina) [inline benchmark/benchmark.js:107]:
```suggestion
    return new Promise((resolve) => {
```

File: benchmark/benchmark.js (hunk 3 above)
Comment (sirenkovladd) [reply benchmark/benchmark.js:106]: I think we can't remove async

https://github.com/tinylibs/tinybench/blob/v2.5.0/src/utils.ts#L31

**Comment by kibertoad** [comment]
> it's written in typescript, which generates code slower than hand-crafted JavaScript.

I still haven't seen any conclusive evidence for that, can you share any?

**Comment by kibertoad** [comment]
> it forces the use of an async function, when a promise is enough.

There is no such thing as "requiring an async function" in TS. All you can do is expect something to return a promise. Whether that is achieved by making function async, or just by returning a promise directly, makes no difference to TS. So it's the usage that is wrong, it can be adjusted to avoid `async`.

**Comment by mcollina** [comment]
> I still haven't seen any conclusive evidence for that, can you share any?

I have not done a scientific study on this, it's more about in-the-field examples.
https://github.com/WebReflection/raytrace provides a good example.

The fundamental problem is that TS can generate quite a bit of added code that's either suboptimal or just useless.

Normally this is not a problem, but for something that _must be as low overhead as possible_, it's a risk I do not want to have.

> There is no such thing as "requiring an async function" in TS. All you can do is expect something to return a promise within an interface. Whether that is achieved by making function async, or just by returning a promise directly, makes no difference to TS. So it's the usage that is wrong, it can be adjusted to avoid async.

They actually force it: https://github.com/tinylibs/tinybench/blob/v2.5.0/src/utils.ts#L31

**Comment by kibertoad** [comment]
> They actually force it: https://github.com/tinylibs/tinybench/blob/v2.5.0/src/utils.ts#L31

Oooof. Yes, that seems unnecesarily restrictive. Are there any better alternatives?

**Comment by Uzlopak** [comment]
I partially worked with @Aslemammad on tinybench v1 (v2 was a complete rewrite by himself). I guess he can fix the async issue. 

On the other hand... Tinybench has, even if ts is adding overhead, still less overhead than benchmark
Js. Simply because it has less deps than benchmark.js. e.g. benchmark js relies heavily on lodash.

Tinybench has the benefit, that when used with 0x or deoptigate your report does not get tainted by lodash and co. This is always super annoying.

**Comment by Uzlopak** [comment]
Lol.

isAsyncFunction looks like code i wrote.

**Review by mcollina — APPROVED** [review]
lgtm, ok

**Comment by sirenkovladd** [comment]
> Oooof. Yes, that seems unnecesarily restrictive. Are there any better alternatives?

https://github.com/tinylibs/tinybench/pull/52

**Comment by sirenkovladd** [comment]
Do you have any blocks to merge PR?

**Comment by kibertoad** [comment]
@sirenkovladd Shouldn't a new version of tinybench be released first? I think the only blocker was the hard requirement for `async`, as long as that is addressed, should be good to go.

**Comment by Uzlopak** [comment]
@sirenkovladd 
Please fix the merge conflicts till next version of tinybench gets released :)

**Comment by sirenkovladd** [comment]
FYI, done

**Comment by Uzlopak** [comment]
Please run a final time the benchmarks and post them here, please

**Comment by Uzlopak** [comment]
@mcollina 

Are you happy :)?

**Comment by Uzlopak** [comment]
wait. you did not fix the remarks by mcollina. so you can now remove the async and return directly the promise. Then again benchmarks please.

**Comment by sirenkovladd** [comment]
```
┌─────────┬──────────────────────────────────┬─────────────┬────────────────────┬───────────┬─────────┐
│ (index) │            Task Name             │   ops/sec   │ Average Time (ns)  │  Margin   │ Samples │
├─────────┼──────────────────────────────────┼─────────────┼────────────────────┼───────────┼─────────┤
│    0    │            'Request'             │  '344,454'  │ 2903.1371215737463 │ '±2.06%'  │ 172228  │
│    1    │         'Custom Request'         │  '26,937'   │ 37122.354565240974 │ '±3.46%'  │  13469  │
│    2    │      'Request With Cookies'      │  '242,493'  │ 4123.821491915785  │ '±1.74%'  │ 121247  │
│    3    │ 'Request With Cookies n payload' │  '225,006'  │ 4444.306685991836  │ '±1.73%'  │ 112504  │
│    4    │            'ParseUrl'            │ '1,095,375' │  912.92936133353   │ '±0.53%'  │ 547688  │
│    5    │       'ParseUrl and query'       │  '127,661'  │ 7833.186162870036  │ '±0.41%'  │  63831  │
│    6    │     'read request body JSON'     │  '88,130'   │ 11346.851367115709 │ '±1.22%'  │  44066  │
│    7    │    'read request body buffer'    │  '120,823'  │ 8276.531522503943  │ '±1.45%'  │  60412  │
│    8    │   'read request body readable'   │  '56,321'   │ 17755.187702620893 │ '±2.74%'  │  28161  │
│    9    │       'Response write end'       │  '28,837'   │ 34676.64944992473  │ '±9.10%'  │  14419  │
│   10    │     'Response writeHead end'     │  '27,481'   │ 36387.79125106221  │ '±27.87%' │  13741  │
│   11    │          'base inject'           │  '18,936'   │ 52807.191164733515 │ '±13.80%' │  9469   │
└─────────┴──────────────────────────────────┴─────────────┴────────────────────┴───────────┴─────────┘
```

**Pull request merged**